### PR TITLE
fix(functions): realign main service script

### PR DESCRIPTION
**Pull request by nyannyacha**
## What kind of change does this PR introduce?

Refactor

## Description

* Resolves a previously intermittent `Status is not defined` issue.
* Enhances Edge-Runtime specific error handling. (BootError, WorkerLimit, WorkerRequestCancelled)
* Make the decorator feature enabled on the user worker (tc39 proposal)
* Add the `onError` callback.

> ~~Note: This PR should be merged when the CLI bumps the edge-runtime image to v1.43.0 or higher.~~

cc @laktek

### CODE REVIEW

File: internal/functions/serve/templates/main.ts [hunk 1]
```diff
@@ -47,8 +76,8 @@ const functionsConfig: Record<string, FunctionConfig> = (() => {
     }
 
     return functionsConfig;
-  } catch (cause) {
-    throw new Error("Failed to parse functions config", { cause });
+  } catch (e) {
```
Comment (laktek) [inline internal/functions/serve/templates/main.ts:79]: nit: do we need to change this?

File: internal/functions/serve/templates/main.ts [hunk 2]
```diff
@@ -69,120 +98,129 @@ async function verifyJWT(jwt: string): Promise<boolean> {
...
+
+    const forceCreate = true;
+    const customModuleRoot = ""; // empty string to allow any local path
+    const cpuTimeSoftLimitMs = 1000;
+    const cpuTimeHardLimitMs = 2000;
+
+    // NOTE(Nyannyacha): Decorator type has been set to tc39 by Lakshan's request,
+    // but in my opinion, we should probably expose this to customers at some
+    // point, as their migration process will not be easy.
+    const decoratorType = "tc39";
```
Comment (laktek) [inline internal/functions/serve/templates/main.ts:159]: yep, let's keep it `tc39` for now and can allow users' to configure it on `config.toml` (not many users use decorators on edge functions yet)

**Review by laktek — APPROVED** [review]

File: internal/functions/serve/templates/main.ts (hunk 1 above)
Comment (nyannyacha) [reply internal/functions/serve/templates/main.ts:79]: It seems to have changed in the process of renaming the variable in the catch statement through the typescript LSP 😓

Comment (nyannyacha) [reply internal/functions/serve/templates/main.ts]: solved

File: internal/functions/serve/templates/main.ts [hunk 3]
```diff
@@ -1,38 +1,67 @@
...
+  STATUS_CODE,
+  STATUS_TEXT,
+} from "https://deno.land/std/http/status.ts";
+
 import * as jose from "https://deno.land/x/jose@v4.13.1/index.ts";
 
-const JWT_SECRET = Deno.env.get("SUPABASE_INTERNAL_JWT_SECRET")!;
-const HOST_PORT = Deno.env.get("SUPABASE_INTERNAL_HOST_PORT")!;
+const SB_SPECIFIC_ERROR_CODE = {
+  BootError: 503, /** Service Unavailable (RFC 7231, 6.6.4) */
```
Comment (sweatybridge) [inline internal/functions/serve/templates/main.ts:9]: Use `STATUS_CODE.XX`?

File: internal/functions/serve/templates/main.ts [hunk 4]
```diff
@@ -69,120 +98,129 @@ async function verifyJWT(jwt: string): Promise<boolean> {
...
+    return getResponse(
+      {
+        code: STATUS_TEXT[STATUS_CODE.InternalServerError],
+        message: "Request failed due to an internal server error",
+        trace: JSON.stringify(e.stack)
+      },
+      STATUS_CODE.InternalServerError
+    )
+  }
+});
```
Comment (sweatybridge) [inline internal/functions/serve/templates/main.ts:226]:
```suggestion
});

```

**Review by sweatybridge — COMMENTED** [review]
nitpicks

**Pull request merged**